osu.Game/Skinning/Editor/SkinSelectionHandler.cs

```diff
@@ -20,6 +20,89 @@ namespace osu.Game.Skinning.Editor
 {
     public class SkinSelectionHandler : SelectionHandler<ISkinnableDrawable>
     {
+        private const string closest_text = @"Closest";
+
+        /// <summary>
+        /// The hash code of the "Closest" menu item in the anchor point context menu.
+        /// </summary>
+        /// <remarks>Needs only be constant and distinct from any <see cref="Anchor"/> values.</remarks>
+        private static readonly int closest_text_hash = closest_text.GetHashCode();
```

Comment:
There is nothing actually guaranteeing that `closest_text.GetHashCode()` doesn't accidentally return a value that clashes with the `Anchor` enum values, except for `GetHashCode()` implementation details - which should not be relied upon. I think that all of this hashing is honestly overdone, and an actual simpler method to do this would be to declare an `enum` (just for creating/displaying the items) or another explicitly-defined mapping.

Reply:
I was experimenting with this before your review, and I can see this being done with a separate menu item instead, pretty much shortened and easier to read:
```diff
diff --git a/osu.Game/Skinning/Editor/SkinSelectionHandler.cs b/osu.Game/Skinning/Editor/SkinSelectionHandler.cs
index f8d41d94fa..783ed6e27f 100644
--- a/osu.Game/Skinning/Editor/SkinSelectionHandler.cs
+++ b/osu.Game/Skinning/Editor/SkinSelectionHandler.cs
@@ -20,41 +20,6 @@ namespace osu.Game.Skinning.Editor
 {
     public class SkinSelectionHandler : SelectionHandler<ISkinnableDrawable>
     {
-        private const string closest_text = @"Closest";
-
-        /// <summary>
-        /// The hash code of the "Closest" menu item in the anchor point context menu.
-        /// </summary>
-        /// <remarks>Needs only be constant and distinct from any <see cref="Anchor"/> values.</remarks>
-        private static readonly int closest_text_hash = closest_text.GetHashCode();
-
-        private static readonly Dictionary<int, string> anchor_menu_items;
-        private static readonly Dictionary<int, string> origin_menu_items;
-
-        static SkinSelectionHandler()
-        {
-            var anchorMenuSubset = new[]
-            {
-                Anchor.TopLeft,
-                Anchor.TopCentre,
-                Anchor.TopRight,
-                Anchor.CentreLeft,
-                Anchor.Centre,
-                Anchor.CentreRight,
-                Anchor.BottomLeft,
-                Anchor.BottomCentre,
-                Anchor.BottomRight,
-            };
-
-            var texts = anchorMenuSubset.Select(a => a.ToString()).Prepend(closest_text).ToArray();
-            var hashes = anchorMenuSubset.Cast<int>().Prepend(closest_text_hash).ToArray();
-
-            var anchorPairs = hashes.Zip(texts, (k, v) => new KeyValuePair<int, string>(k, v)).ToArray();
-            anchor_menu_items = new Dictionary<int, string>(anchorPairs);
-            var originPairs = anchorPairs.Where(pair => pair.Key != closest_text_hash);
-            origin_menu_items = new Dictionary<int, string>(originPairs);
-        }
-
         private Anchor getClosestAnchorForDrawable(Drawable drawable)
         {
             var parent = drawable.Parent;
@@ -267,35 +232,49 @@ protected override void OnSelectionChanged()

         protected override IEnumerable<MenuItem> GetContextMenuItemsForSelection(IEnumerable<SelectionBlueprint<ISkinnableDrawable>> selection)
         {
-            int checkAnchor(Drawable drawable) =>
-                drawable.UsingClosestAnchor().Value
-                    ? closest_text_hash
-                    : (int)drawable.Anchor;
-
             yield return new OsuMenuItem(nameof(Anchor))
             {
-                Items = createAnchorItems(anchor_menu_items, checkAnchor, applyAnchor).ToArray()
+                Items = new[]
+                {
+                    new TernaryStateRadioMenuItem("Closest", MenuItemType.Standard, _ => applyClosestAnchor())
+                    {
+                        State = { Value = GetStateFromSelection(selection, c => ((Drawable)c.Item).UsingClosestAnchor().Value) }
+                    },
+                }.Concat(createAnchorItems((d, a) => d.Anchor == a && !d.UsingClosestAnchor().Value, applyCustomAnchor)).ToList()
             };

             yield return new OsuMenuItem(nameof(Origin))
             {
                 // Origins can't be "closest" so we just cast to int
-                Items = createAnchorItems(origin_menu_items, d => (int)d.Origin, applyOrigin).ToArray()
+                Items = createAnchorItems((d, o) => d.Origin == o, applyOrigin).ToArray()
             };

             foreach (var item in base.GetContextMenuItemsForSelection(selection))
                 yield return item;

-            IEnumerable<TernaryStateMenuItem> createAnchorItems(IDictionary<int, string> items, Func<Drawable, int> checkFunction, Action<int> applyFunction
) =>
-                items.Select(pair =>
+            IEnumerable<TernaryStateMenuItem> createAnchorItems(Func<Drawable, Anchor, bool> checkFunction, Action<Anchor> applyFunction)
+            {
+                var displayableAnchors = new[]
                 {
-                    var (hash, text) = pair;
-
-                    return new TernaryStateRadioMenuItem(text, MenuItemType.Standard, _ => applyFunction(hash))
+                    Anchor.TopLeft,
+                    Anchor.TopCentre,
+                    Anchor.TopRight,
+                    Anchor.CentreLeft,
+                    Anchor.Centre,
+                    Anchor.CentreRight,
+                    Anchor.BottomLeft,
+                    Anchor.BottomCentre,
+                    Anchor.BottomRight,
+                };
+
+                return displayableAnchors.Select(a =>
+                {
+                    return new TernaryStateRadioMenuItem(a.ToString(), MenuItemType.Standard, _ => applyFunction(a))
                     {
-                        State = { Value = GetStateFromSelection(selection, c => checkFunction((Drawable)c.Item) == hash) }
+                        State = { Value = GetStateFromSelection(selection, c => checkFunction((Drawable)c.Item, a)) }
                     };
                 });
+            }
         }

         private static void updateDrawablePosition(Drawable drawable, Vector2 screenSpacePosition)
@@ -304,19 +283,15 @@ private static void updateDrawablePosition(Drawable drawable, Vector2 screenSpac
                 drawable.Parent.ToLocalSpace(screenSpacePosition) - drawable.AnchorPosition;
         }

-        private void applyOrigin(int hash)
+        private void applyOrigin(Anchor origin)
         {
-            var anchor = (Anchor)hash;
-
             foreach (var item in SelectedItems)
             {
                 var drawable = (Drawable)item;

                 var previousOrigin = drawable.OriginPosition;
-                drawable.Origin = anchor;
+                drawable.Origin = origin;
                 drawable.Position += drawable.OriginPosition - previousOrigin;
-
-                updateDrawableAnchorIfUsingClosest(drawable);
             }
         }

@@ -327,39 +302,35 @@ private void applyOrigin(int hash)
         private Quad getSelectionQuad() =>
             GetSurroundingQuad(SelectedBlueprints.SelectMany(b => b.Item.ScreenSpaceDrawQuad.GetVertices().ToArray()));

-        private void applyAnchor(int hash)
+        private void applyClosestAnchor()
         {
             foreach (var item in SelectedItems)
             {
                 var drawable = (Drawable)item;

-                var anchor = getAnchorFromHashAndDrawableAndRecordWhetherUsingClosestAnchor(hash, drawable);
+                drawable.UsingClosestAnchor().Value = true;
+                updateDrawableAnchor(drawable, getClosestAnchorForDrawable(drawable));
+            }
+        }
+
+        private void applyCustomAnchor(Anchor anchor)
+        {
+            foreach (var item in SelectedItems)
+            {
+                var drawable = (Drawable)item;

+                drawable.UsingClosestAnchor().Value = false;
                 updateDrawableAnchor(drawable, anchor);
             }
         }

-        private static void updateDrawableAnchor(Drawable drawable, Anchor anchor)
+        private void updateDrawableAnchor(Drawable drawable, Anchor anchor)
         {
             var previousAnchor = drawable.AnchorPosition;
             drawable.Anchor = anchor;
             drawable.Position -= drawable.AnchorPosition - previousAnchor;
         }

-        private Anchor getAnchorFromHashAndDrawableAndRecordWhetherUsingClosestAnchor(int hash, Drawable drawable)
-        {
-            var isUsingClosestAnchor = drawable.UsingClosestAnchor();
-
-            if (hash == closest_text_hash)
-            {
-                isUsingClosestAnchor.Value = true;
-                return getClosestAnchorForDrawable(drawable);
-            }
-
-            isUsingClosestAnchor.Value = false;
-            return (Anchor)hash;
-        }
-
         private static void adjustScaleFromAnchor(ref Vector2 scale, Anchor reference)
         {
             // cancel out scale in axes we don't care about (based on which drag handle was used).
```

Reply:
> There is nothing actually guaranteeing that `closest_text.GetHashCode()` doesn't accidentally return a value that clashes with the `Anchor` enum values, except for `GetHashCode()` implementation details - which should not be relied upon.

This can be solved with a static computation (something simple like `~(Anchor.x0 | Anchor.x1 | Anchor.x2 | Anchor.y0 | Anchor.y1 | Anchor.y2)` would meet the requirements).

> I think that all of this hashing is honestly overdone, and an actual simpler method to do this would be to declare an `enum` (just for creating/displaying the items) or another explicitly-defined mapping.

So... I tried this, and it's a ton of boilerplate and more error-prone. I think one hash plus an existing `enum` is more elegant, but let me know what you think.

I could finish a working implementation but I think this is enough of an illustration:
```cs
private enum MenuAnchor
{
    Closest,
    TopLeft,
    TopCentre,
    TopRight,
    CentreLeft,
    Centre,
    CentreRight,
    BottomLeft,
    BottomCentre,
    BottomRight
}

private static MenuAnchor getMenuAnchorForDrawable(Drawable drawable)
{
    if (drawable.UsingClosestAnchor().Value)
        return MenuAnchor.Closest;

    const Anchor mask = Anchor.x0 | Anchor.x1 | Anchor.x2 | Anchor.y0 | Anchor.y1 | Anchor.y2;

    switch (drawable.Anchor & mask)
    {
        case Anchor.TopLeft:
            return MenuAnchor.TopLeft;

        case Anchor.TopCentre:
            return MenuAnchor.TopCentre;

        case Anchor.TopRight:
            return MenuAnchor.TopRight;

        case Anchor.CentreLeft:
            return MenuAnchor.CentreLeft;

        case Anchor.Centre:
            return MenuAnchor.Centre;

        case Anchor.CentreRight:
            return MenuAnchor.CentreRight;

        case Anchor.BottomLeft:
            return MenuAnchor.BottomLeft;

        case Anchor.BottomCentre:
            return MenuAnchor.BottomCentre;

        case Anchor.BottomRight:
            return MenuAnchor.BottomRight;

        default:
            throw new ArgumentOutOfRangeException(nameof(drawable), @"Drawable.Anchor is invalid.");
    }
}
```

An alternative implementation would be one where `MenuAnchor` copies the flags from `Anchor`, but at that point it's just a cast anyway.

Reply:
> I was experimenting with this before your review, and I can see this being done with a separate menu item instead, pretty much shortened and easier to read

I have this saved as a stash but I want to resolve the `enum` question first. I think the approach is fundamentally more sound than what I had, though.

Reply:
If I'm honest, I prefer boilerplate to unnecessary complexity, and the hashing stuff firmly places in the second camp to me. I haven't read @frenzibyte's suggestion overly thoroughly but it seemed more tolerable for sure.

Reply:
Okay, I'll proceed in that direction.